docs/adrs/0001-support-caching-deps-for-monorepos.md

```diff
@@ -0,0 +1,48 @@
+# 0. Support caching dependencies for mono repos
+Date: 2021-07-13
+
+Status: Proposed
+
+## Context
+Currently, `actions/setup-node` supports caching dependencies for Npm and Yarn package managers.  
+For the first iteration, we have decided to not support cases where `package-lock.json` / `yarn.lock` are located outside of repository root.  
+Current implementation searches the following file patterns in the repository root: `package-lock.json`, `yarn.lock` (in order of resolving priorities)
+
+Obviously, it made build-in caching unusable for mono-repos and repos with complex structure.  
+We would like to revisit this decision and add customization for dependencies lock file location.
+
+## Proposal
+We have the following options:
+1. Allow to specify  directory where `package-lock.json` or `yarn.lock` are located
+2. Allow to specify path to the dependencies lock file (including directory path and filename)
+
+The second option looks more generic because it allows to:
+- specify multiple dependencies files using file patterns like `**/package-lock.json` ([one of recommended approaches in actions/cache](https://github.com/actions/cache/blob/main/examples.md#macos-and-ubuntu))
+- specify custom dependencies files like `src/npm-shrinkwrap.json`
+- change default resolving priority if both `yarn.lock` and `package-lock.json` exist in repository
+
+## Decision
+
+Add `package-lock-file` input that will accept path (relative to repository root) to dependencies lock file.  
+If provided path contains wildcards, the action will search all maching files and calculate common hash like `${{ hashFiles('**/package-lock.json') }}` YAML construction does.  
+The hash of provided matched files will be used as a part of cache key.
+
+Yaml examples:
+```yml
+steps:
+- uses: actions/checkout@v2
+- uses: actions/setup-node@v2
+  with:
+    node-version: 14
+    cache: npm
+    package-lock-file: 'sub-project/package-lock.json'
```

Comment:
Should we use a more generic name such as 'cache-path' so that we can use the same for other setup-actions? Also, we probably don't need apostrophes here.

Reply:
I think `cache-path` can confuse users because it sounds like a path to cache or folder that will be cached. Actually, this variable points to dependencies lock file (file that describes / define dependencies of the project). IMHO, `package-lock-file` or `package-lock-path` is generic enough to be used across other actions too. Another option - something like `deps-lock-path`.

Reply:
Personally I'm not a fan of `cache-path` for the same reasons.

My vote goes to this input being called `cache-dependency-path`.
 - It's prefixed with `cache` so it's immediately obvious what it's for. It will also make the YAML more readable with 2 `cache-*` input keys right next to each other.
- Since we will be supporting wildcards for multiple paths, it would be confusing to have something with a singular `*-file` in the name be used to cache multiple file**s**. This observation generally leads me to prefer having `*-path` in the name vs `*-file` or `*-files`.
- Generic enough where it can be used in other actions. For node we're dealing with lock files, but to be as generic as possible we should avoid having `lock` in the name since other first party actions like Python have different dependency mechanisms. `dependency` in the input key feels like a good middle ground that will make this easily reusable across all first party actions

Reply:
I agree with Konrad arguments above.

Reply:
+1 agreed on points above